samsung-galaxybook.c

```diff
@@ -575,6 +581,64 @@ static ssize_t battery_saver_percent_show(struct device *dev, struct device_attr
 
 static DEVICE_ATTR_RW(battery_saver_percent);
 
+static ssize_t charge_control_end_threshold_store(struct device *dev, struct device_attribute *attr,
+				const char *buffer, size_t count)
+{
+	u8 value;
+	int err;
+
+	if (!count || kstrtou8(buffer, 0, &value))
+		return -EINVAL;
+
+	// 100 is a special value that means "off"
+	if (value == 100)
+		value = 0;
+
+	err = battery_saver_percent_acpi_set(galaxybook_ptr, value);
+	if (err)
+		return err;
+
+	return count;
+}
+
+static ssize_t charge_control_end_threshold_show(struct device *dev, struct device_attribute *attr,
+				char *buffer)
+{
+	u8 value;
+	int err;
+
+	err = battery_saver_percent_acpi_get(galaxybook_ptr, &value);
+	if (err)
+		return err;
+
+	if (value == 0)
+		value = 100;
+
+	return sysfs_emit(buffer, "%d\n", value);}
+
+static DEVICE_ATTR_RW(charge_control_end_threshold);
+
+static int samsung_battery_add(struct power_supply *battery, struct acpi_battery_hook *hook)
+{
+	if (device_create_file(&battery->dev,
+			       &dev_attr_charge_control_end_threshold))
+		return -ENODEV;
+
+	return 0;
+}
+
+static int samsung_battery_remove(struct power_supply *battery, struct acpi_battery_hook *hook)
+{
+	device_remove_file(&battery->dev,
+			   &dev_attr_charge_control_end_threshold);
+	return 0;
+}
+
+static struct acpi_battery_hook battery_hook = {
+	.add_battery = samsung_battery_add,
+	.remove_battery = samsung_battery_remove,
+	.name = "Samsung Galaxy Book Battery Extension",
+};
```

Comment:
Should we remove the previous way to set the battery threshold charge?

Reply:
This is super cool and very glad you found this -- yes my take is that the other one should be removed. If you want, we can leave your PR as-is and then fix in the branch somehow before merging to main, or otherwise feel free to remove it here in your branch/PR. Feel free to say if you want to go ahead or just leave it and I can fix it after merging :)

Also this had me re-thinking a bit of my initial idea of "if someone is using this device and knows what things are called in Windows, maybe it should be roughly named the same in Linux, so it is easy for them to understand"
But now I am re-thinking this with this finding (plus using platform profiles instead of a custom "performance_mode" interface is another example of this).  So my thinking was to try and look through existing kernel code and try to find anywhere else where any other of these features might have a more "standard" way of handling or a future "standard" way have handling and will make some small adjustments based on that as well.

One example: thinking to change the attribute "usb_charging" to "usb_charge" as this name is already relatively widely used for the same feature in the Linux kernel by other devices. Still looking for other opportunities like this but this is a really good start, I think!